### ISSUE TITLE
[iOS] Feature request: periodically cycle through servers in selected geography (optional setting)

### ISSUE THREAD

**Issue by M09NST3R** [mullvad/mullvadvpn-app#2623]
## Issue report

Operating system: iOS (applies to other operating systems as well, but on these other systems there's typically a way to easily script the desired behavior, e.g., via Cron jobs)

App version: any

### Issue description

Would be great if there was a setting in the iOS app to randomly and automatically change the VPN server every X hours/minutes. Otherwise, it (unnecessarily) becomes easier to de-anonymize users via metadata correlation.

**Comment by faern** [comment]
Thank you for the suggestion. We don't have any existing plans for this type of functionality. But it's an interesting feature. We currently rotate the WireGuard key every four days on iOS. And this cause a reconnect. So we kind of already have this functionality, but it's hardcoded to four days.

What intervals did you have in mind for such a feature. How long interval would you set? You say "X hours/minutes". I interpret this as you wanting to rotate pretty often. That often will be really bad for connection stability. Even at one hour I'd imagine this would impact usability noticeably.

**Comment by M09NST3R** [comment]
> Thank you for the suggestion. We don't have any existing plans for this type of functionality. But it's an interesting feature. We currently rotate the WireGuard key every four days on iOS. And this cause a reconnect. So we kind of already have this functionality, but it's hardcoded to four days.
> 
> What intervals did you have in mind for such a feature. How long interval would you set? You say "X hours/minutes". I interpret this as you wanting to rotate pretty often. That often will be really bad for connection stability. Even at one hour I'd imagine this would impact usability noticeably.

On desktop I have a cron job that changes the connection randomly every 30-60 minutes. It checks for the existence of a dotfile in my home folder and if it can't find it, it won't do the reconnect. This allows me to disable the functionality quickly during video calls etc. (`rm ~/.dotfile`).

On mobile, we would probably want a widget and/or an easy-to-access switch in the app itself.

**Comment by xanoni** [comment]
FYI, I increased the interval from 30-60 minutes as it was causing too many interruptions during video calls. I now randomly cycle through servers every 2-4 hours instead. If I touch a file in my home folder, the script recognizes that and skips for one cycle.

**Comment by faern** [comment]
This is mainly why we don't have said feature. It will cause a lot of interruption for users and it's going to be non-trivial to explain how it works and how to use it. Along with the extra buttons and controls needed to temporarily pause it and all.

**Comment by xanoni** [comment]
Ok sad panda. I think it should be less of a problem on mobile, though, given people don't do video conferencing as often (at least I don't) and phones are designed to deal with network changes. And mobile is also where I'm struggling, as it's more difficult to script the desired behavior.

I've read about two ways other people solve this:
1) Round robin DNS server that references changing Mullvad VPN servers
2) Set up VPS with multiple network interfaces and then change the mapping of incoming (client) to outgoing (VPN) interfaces regularly

These certainly work, but involve trade-offs and work. Seems unnecessary given it could easily be done client-side.

**Comment by RequestPrivacy** [comment]
I would love such a feature and think it's worth a thorough discussion. Currently I'm doing this manually. @xanoni the request would also be a nice one for all clients, not just for iOS. Maybe update the title to broaden the scope to Android/Linux/Windows etc.?

> How long interval would you set? You say "X hours/minutes". I interpret this as you wanting to rotate pretty often. That often will be really bad for connection stability

For me a handy cycle time would be measured in days, not hours. So it would be okay if the switch is made every time the device reconnects to a wifi (e.g. after a new start up/switching online from airplane mode/coming home from work/connecting to a public wifi etc.). To prevent the trigger from cycling through servers "to often", one could imagine a max. per day limit?!

To users this is a an easy and simple condition and should occur once in a wile but also not to often in normal usage. The other questions is if there is a real privacy benefit?

**Comment by xanoni** [comment]
> The other questions is if there is a real privacy benefit?

Well, you are hiding behind a pool of other users, but afaik there are no statistics published by Mullvad how big this pool is (i.e., how many concurrent user connections per VPN server). The longer you stay behind the same server, the easier it becomes to identify you statistically. (Or, at least to know that you are NOT a certain user.)

**Comment by 6e6f61** [comment]
Is there any chance of movement on this? I think it'd be really neat. And as @RequestPrivacy mentioned, rotating the server whenever internet is disconnected shouldn't cause any disruptions.

**Comment by xanoni** [comment]
Also still interested.

**Comment by nkud90** [comment]
I have found my personal server with port forwarded via MullVad VPN and Apache running not accessible after about one week and a half of good operation. 
Manual reconnection solved the issue.
Therefore I am also interested in the auto-reconnect every "n" days feature.

(Thank you for your job, MullVad VPN is awesome).

**Comment by lordz0rn** [comment]
Hey there,

+1.
My scenario is that i‘m on the road with an ipad.
(wg automatically connects on mobile and foreign wifi, so it‘s practically connected all the time)
The wg server sits behind a daily changing public IP.
(pretty common here in medieval germany)
So every morning the iPad is offline and I have to manually force a wg reconnect.
(de- and reactivating wifi - and even flight mode - does not trigger a reconnect to the new public IP)

Any kind of mitigation to this would be appreciated :)
Auto reconnect after 24h, a Widget, forced server-public-ip-dns-check or shortcut-support - anything would help.


Actual workaround: no wg auto-connect on mobile, so de- and reactivating wifi triggers a full reconnect… :)


Thanks again. wireguard does a great job in many ways :)

**Comment by faern** [comment]
@lordz0rn 

>  The wg server sits behind a daily changing public IP.

Do you have your own WireGuard server? Well, then you can't use our app. Our iOS app only works against our Mullvad VPN servers. So I'm confused here around what you are trying to achieve. Our servers do not change public IP addresses. At least not that often. Also, our iOS app automatically reconnects when it detects that the tunnel is broken. This issue is not about reconnecting to fix broken tunnels, but to reconnect to improve privacy, which is a different use case.

**Comment by lordz0rn** [comment]
yep. dang. sorry. must have been benighted for I posted in the completely wrong github forum :) … sorry again.

**Comment by zbarnz** [comment]
I would be interested in this feature. I have a webscraper running on one of my machines and it would be nice to just cycle through IP's using mulladVPN (which I already pay for) instead of setting up some kind of rotating proxy.

**Comment by MagicLike** [comment]
This feature could also be coupled with the custom lists from the last update, that only the servers/locations on the list get rotated through.
